js/libs/keycloak-js/src/keycloak.js

```diff
@@ -752,9 +753,13 @@ function Keycloak (config) {
 
         if (error) {
             if (prompt != 'none') {
-                var errorData = { error: error, error_description: oauth.error_description };
-                kc.onAuthError && kc.onAuthError(errorData);
-                promise && promise.setError(errorData);
+                if (oauth.error_description && oauth.error_description === "Authentication+expired.+Please+try+again.") {
+                    kc.login(oauth.loginOptions);
```

Comment:
This seems like it could cause infinite recursion. Is there no better way to do this than to match the description?

Reply:
@jonkoops Thanks for this point and for the review!

I am not seeing any better way as `error` and `error_description` are parameters sent from Keycloak to signal to the client what happened and this particular error means that client can/should retry authentication (as user is already logged-in and has SSO session, but authentication session is expired/removed. Hence server is not able to properly finish authentication and so needs to ask client to "retry authentication" (BTW. added some additional description to this PR).

ATM I am not seeing how infinite loop can happen as server redirects this error to client just when authenticationSession is expired/removed, but when client sends another login request to KEycloak, it will new authenticationSession, which is then able to properly finish authentication flow (or keep display login screen if something like `prompt=login` is used etc). So unless there is some bug on server, infinite redirect should not happen.

But for sake of avoid "infinite redirect", it is maybe possible to add some flag to `oauthState` on keycloak.js side when retried login request is sent, so if server redirects back with same error (which it should not), the keycloak.js may not try retry login again.

I am not sure if it is needed as IMO it adds some additional complexity to `keycloak.js` with very little added value, but if you prefer, I can try to add something like this?

Reply:
> I am not sure if it is needed as IMO it adds some additional complexity to keycloak.js with very little added value, but if you prefer, I can try to add something like this?

No, I think your motivation here is fine. My main concern is matching human-readable messages that are likely so change and then break this logic. Are you sure there is not some kind of error code/enum we can match? What are the contents of `error`?

Reply:
@jonkoops Yes, that is great point. I was also thinking about that.

In summary, it works like:

- The `error` is supposed to be the the "known code" like for instance `server_error` or `invalid_request`. The list of available error codes is available in the OAuth2 specification here: https://datatracker.ietf.org/doc/html/rfc6749#page-45 . There are few available error codes in the specification and none of them is perfect for tell the client that "authentication expired" . We discussed with Stian that best code from those available in the specs might be `temporarily_unavailable` 

- The `error_description` is supposed to be human-readable text. And it is not prescribed in the specification what exactly should be here and hence it is possible to add some custom text. So for this, we have the options like:
  - Use human-readable text like `Authentication+expired.+Please+try+again.`
  - Use something shorter and without spaces like `authentication_expired` , which is not exactly human-readable text, but has a smaller chance that it will change and does not contains spaces (which can be problematic).

I was like 50 / 50 which of those options to use. I chosen the human-readable option, but based on your feedback (and also based on the fact that clients may need to check this code), I am more inclined to update this.

@jonkoops @rmartinc @graziang Anyone has any issue with changing `error_description` to something like `authentication_expired` ?

Reply:
The message `authentication_expired` is OK to me too.